src/scales/scale.category.js

```diff
@@ -49,10 +50,19 @@ module.exports = Scale.extend({
 
 	getLabelForIndex: function(index, datasetIndex) {
 		var me = this;
-		var data = me.chart.data;
+		var chart = me.chart;
+		var data = chart.data;
 		var isHorizontal = me.isHorizontal();
+		var controller = chart.getDatasetMeta(datasetIndex).controller;
 
-		if (data.yLabels && !isHorizontal) {
+		// For controllers supporting getValueScaleId, we can be sure if this is a value scale.
+		// For object data ({x,y} etc.) we can rely on getRightValue to return correct thing.
+		// For others, check if data is a label and use that instead (so this is an value index, but values are labels)
+		var isValueScale = controller.getValueScaleId
+			? controller.getValueScaleId() === me.id
+			: helpers.isObject(data[0]) || (!isHorizontal && me.getLabels().indexOf(data.datasets[datasetIndex].data[index]) !== -1);
```

Comment:
you could make a variable for `data.datasets[datasetIndex].data[index]` since it's used multiple times

Reply:
Actually though, this seems potentially not right. What if it finds the label at the wrong location?

Maybe it needs to be something more like `me.getLabels()[index] === data.datasets[datasetIndex].data[index]`?

Reply:
Say you have 2 labels (like in that [pen](https://codepen.io/kurkle/pen/bOepaQ)). And we are requesting label for index `5`.  `me.getlabels()[index]` would be out of bounds.

The added `index < data.length` is for the opposite case (eg. 5 labels, but only 2 data points).